### ISSUE TITLE
Scratchpads do not work properly for emacsclient

### ISSUE THREAD

**Issue by nmccarty** [hyprland-community/pyprland#8]
Reopen of #6 with more information
Video demonstration: https://youtu.be/chP2NtZRRVw (youtube is still processing the video at time of submitting, give it a few minutes if high resolution isn't available yet) 

The JSON errors were from a copy/paste error extracting just one section from my config (and I did actually try using a wrapper script in my config)

The window is set floating and has the height correctly configured in the actual config, the resulting window doesn't appear as tiled, and, well, never actually appears at all despite actually getting launched

Unfortunately, simply replacing the command with `emacs` doesn't result in something equivlant, `emacs` and `emacsclient` are entirely different commands that have very different semantics. The `emacsclient` command creates a new window and attaches it to an existing emacs session, where as `emacs` always creates a new session. 

The `emacsclient` command I am using runs fine from the terminal, and produces a floating window with the expected geometry. 

Here is the current complete state of my pyprland config:
```json
  {
    "pyprland": {
      "plugins": [
        "scratchpads"
      ]
    },
    "scratchpads": {
      "term": {
        "command": "alacritty --class alacritty-dropterm --working-directory ~ -e tmux new -A -s scratch",
        "animation": ""
      },
      "emacs": {
        "command": "/nix/store/61alz9sbbbnysgq8az0snc6pr3y2lldb-emacsStart.sh",
        "animation": "",
        "unfocus": "hide"
      },
      "ario": {
        "command": "ario",
        "animation": ""
       },
      "spotify": {
        "command": "spotify",
        "animation": ""
       }
    }
  }
```

The contents of the included script:
```sh
  #!/nix/store/8fv91097mbh5049i9rglc73dx6kjg3qk-bash-5.2-p15/bin/bash
  exec emacsclient -c --frame-parameters='(quote (name . "scratchmacs"))'
```
(tested again and calling this script from the terminal creates a floating window with the expected geometry)

and the output from `DEBUG=1 pypr`, popping open and closing the terminal scratchpad to make sure pypr was working, and then attempting to open the emacs scratchpad:
```
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish

~
01:03 PM❯ set -x DEBUG 1

~
01:03 PM❯ pypr
EVT event_createworkspace(special)
EVT event_activewindow(Alacritty,pypr ~)
EVT event_activewindowv2(1ecf400)
EVT event_openwindow(1d2d600,special,emacs,scratchmacs)
(JS)>>> clients
EVT event_activewindow(Alacritty,pypr ~)
EVT event_activewindowv2(1ecf400)
EVT event_openwindow(1ecdfa0,special,alacritty-dropterm,Alacritty)
(JS)>>> clients
>>> movetoworkspacesilent special:scratch_term,address:0x1ecdfa0
<<< b'ok'
EVT event_createworkspace(special:scratch_term)
EVT event_movewindow(1ecdfa0,special:scratch_term)
EVT event_activewindow(Alacritty,pypr ~)
EVT event_activewindowv2(1ecf400)
EVT event_activewindow(Alacritty,pypr ~)
EVT event_activewindowv2(1ecf400)
EVT event_openwindow(1dbda00,special,ario,Ario)
(JS)>>> clients
>>> movetoworkspacesilent special:scratch_ario,address:0x1dbda00
<<< b'ok'
EVT event_createworkspace(special:scratch_ario)
EVT event_movewindow(1dbda00,special:scratch_ario)
EVT event_activewindow(Alacritty,pypr ~)
EVT event_activewindowv2(1ecf400)
EVT event_windowtitle(1dbda00)
EVT event_activewindow(Alacritty,pypr ~)
EVT event_activewindowv2(1ecf400)
EVT event_openwindow(20dbfd0,special,,Spotify)
(JS)>>> clients
>>> movetoworkspacesilent special:scratch_spotify,address:0x20dbfd0
<<< b'ok'
EVT event_createworkspace(special:scratch_spotify)
EVT event_movewindow(20dbfd0,special:scratch_spotify)
EVT event_activewindow(Alacritty,pypr ~)
EVT event_activewindowv2(1ecf400)
CMD: run_toggle(['term'])
(JS)>>> activewindow
(JS)>>> monitors
(JS)>>> clients
>>> moveworkspacetomonitor special:scratch_term DP-3
<<< b'ok'
>>> movetoworkspacesilent 1,address:0x1ecdfa0
EVT event_movewindow(1ecdfa0,1)
EVT event_activewindow(alacritty-dropterm,Alacritty)
EVT event_activewindowv2(1ecdfa0)
<<< b'ok'
>>> focuswindow address:0x1ecdfa0
EVT event_activewindow(alacritty-dropterm,Alacritty)
EVT event_activewindowv2(1ecdfa0)
<<< b'ok'
EVT event_activewindow(alacritty-dropterm,Alacritty)
EVT event_activewindowv2(1ecdfa0)
EVT event_destroyworkspace(special:scratch_term)
CMD: run_toggle(['term'])
>>> movetoworkspacesilent special:scratch_term,address:0x1ecdfa0
EVT event_createworkspace(special:scratch_term)
EVT event_movewindow(1ecdfa0,special:scratch_term)
EVT event_activewindow(Alacritty,pypr ~)
EVT event_activewindowv2(1ecf400)
<<< b'ok'
CMD: run_toggle(['emacs'])
(JS)>>> activewindow
(JS)>>> monitors
(JS)>>> clients
scratchpads::run_toggle(('emacs',)) failed:
Traceback (most recent call last):
  File "/nix/store/ivdglgzgnkwqk9ky7z3rl7d2vhk0b1v1-python3.10-pyprland-1.3.1/lib/python3.10/site-packages/pyprland/command.py", line 51, in _callHandler
    await getattr(plugin, full_name)(*params)
  File "/nix/store/ivdglgzgnkwqk9ky7z3rl7d2vhk0b1v1-python3.10-pyprland-1.3.1/lib/python3.10/site-packages/pyprland/plugins/scratchpads.py", line 209, in run_toggle
    await self.run_show(uid)
  File "/nix/store/ivdglgzgnkwqk9ky7z3rl7d2vhk0b1v1-python3.10-pyprland-1.3.1/lib/python3.10/site-packages/pyprland/plugins/scratchpads.py", line 306, in run_show
    await self.updateScratchInfo(item)
  File "/nix/store/ivdglgzgnkwqk9ky7z3rl7d2vhk0b1v1-python3.10-pyprland-1.3.1/lib/python3.10/site-packages/pyprland/plugins/scratchpads.py", line 226, in updateScratchInfo
    await scratch.updateClientInfo()
  File "/nix/store/ivdglgzgnkwqk9ky7z3rl7d2vhk0b1v1-python3.10-pyprland-1.3.1/lib/python3.10/site-packages/pyprland/plugins/scratchpads.py", line 105, in updateClientInfo
    assert isinstance(clientInfo, dict)
AssertionError
```

After another quick round of troubleshooting, it looks like pypr is indeed actually creating the emacsclient window, hyprctl clients shows it there in it's special workspace, pypr is just struggling with calling it back from the special workspace (see video demonstration)

**Comment by fdev31** [comment]
Just a wild guess, are you able to test such patch ?
```diff
diff --git a/pyprland/plugins/scratchpads.py b/pyprland/plugins/scratchpads.py
index 9118f18..09e5233 100644
--- a/pyprland/plugins/scratchpads.py
+++ b/pyprland/plugins/scratchpads.py
@@ -305,6 +305,7 @@ class Extension(Plugin):
             self.start_scratch_command(uid)
             while uid in self._respawned_scratches:
                 await asyncio.sleep(0.05)
+            await asyncio.sleep(2)
 
         item.visible = True
         monitor = await get_focused_monitor_props()
```

**Comment by nmccarty** [comment]
Ye, I'll give it a spin in a bit

**Comment by fdev31** [comment]
Ok, this is not the race condition I had in mind. The problem is completely different:

Pypr starts & manage the processes, from the PID of each it can figure which client window corresponds.

In the case of Emacs with the client command, the client is created by the daemon, the PID of the client process is never known by hyprland. On the other side, the PID of the daemon process isn't known by pyprland.

I may need to think more about this since it doesn't really fit in the current program architecture, but I can see (not so elegant) ways to work around this, maybe you have more ideas, or just a preference to express:

- add some `external_pid` option, reading the PID from a file instead of relying on the PID of the process
- add some `daemon_command`, similar to `command` but would not restart automatically - the PID of this command would be considered to "find" the client window in case none is found having the command's PID 
- add some `class` option, which would be used as a fallback in case there is no matching PID (that's how primitive versions of the scratchpad were working, I removed that since it was causing ambiguities in some scenarios)

**Comment by fdev31** [comment]
Try pulling the latest git version and use the following in your emacs scratchpad:

```
      "lazy": true,
      "class": "Emacs"
```

I get some strange results without using the lazy option, but with it all seems fine...

**Comment by fdev31** [comment]
I released the `1.4.0` containing support for class match.
It's not documented yet: I'm waiting for your feedback.

**Comment by fdev31** [comment]
Thanks, I added a page about this: https://github.com/hyprland-community/pyprland/wiki/Troubleshooting
